core/server/master/src/main/java/alluxio/master/meta/JournalSpaceMonitor.java

```diff
@@ -0,0 +1,176 @@
+    // command allowed by the POSIX specification
+    CommandReturn output = getRawDiskInfo();
+    if (output.getExitCode() != 0) {
+      throw new IOException("Command failed with exit code " + output.getExitCode());
+    }
+    List<JournalDiskInfo> infos = Arrays.stream(output.getOutput().split("\n"))
+        .skip(1)
+        .map(infoLine -> Arrays.stream(infoLine.split(" "))
+              .map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList()))
+        .filter(data -> data.size() >= 6)
+        .map(data -> {
+          String diskPath = data.get(0);
+          // total size is not necessarily (used size + available size)
+          long totalSize = Long.parseLong(data.get(1));
+          long usedSize = Long.parseLong(data.get(2));
+          long availableSize = Long.parseLong(data.get(3));
+          String mountedFs = data.get(5);
+          return new JournalDiskInfo(diskPath, totalSize, usedSize, availableSize, mountedFs);
+        }).collect(Collectors.toList());
+    mMetricInfo.set(infos.stream().collect(Collectors.toMap(JournalDiskInfo::getDiskPath, f -> f)));
+    infos.forEach(info -> {
+      MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+              MetricKey.MASTER_JOURNAL_SPACE_FREE_BYTES.getName()
+                  + "Device" + MetricsSystem.escape(new AlluxioURI(info.getDiskPath()))), () -> {
+          JournalDiskInfo metricInfo = mMetricInfo.get().get(info.getDiskPath());
+          if (metricInfo != null) {
+            return metricInfo.getAvailableBytes();
+          } else {
+            return -1;
+          }
+        });
+      MetricsSystem.registerGaugeIfAbsent(MetricsSystem.getMetricName(
+          MetricKey.MASTER_JOURNAL_SPACE_FREE_PERCENT.getName()
+              + "Device" + MetricsSystem.escape(new AlluxioURI(info.getDiskPath()))), () -> {
+          JournalDiskInfo metricInfo = mMetricInfo.get().get(info.getDiskPath());
+          if (metricInfo != null) {
+            return metricInfo.getPercentAvailable();
+          } else {
+            return -1;
+          }
+        });
+    });
+    return infos;
+  }
+
+  /**
+   * @return a list warning messages that can be logged or displayed to the user in the UI
+   */
+  public List<String> getJournalDiskWarnings() {
+    try {
+      List<JournalDiskInfo> currentInfo = getDiskInfo();
+      return currentInfo.stream()
+          .filter(info -> info.getPercentAvailable() < mWarnCapacityPercentThreshold)
+          .map(info -> String.format(
+                  "The journal disk %s backing the journal has only %.2f%% space left - %s",
+              info.getDiskPath(), info.getPercentAvailable(), info))
+          .collect(Collectors.toList());
+    } catch (IOException e) {
+      LogUtils.warnWithException(LOG, "Failed to get journal disk information. Critical warnings "
+          + "about journal disk space may not appear in the logs.", e);
+    }
+    return Collections.emptyList();
```

Comment:
Should we include the "failed to get journal disk info" warning in the returned warning list?

Reply:
I am personally not so sure. If returned, the warning would show up in the UI which might indicate to users there is a problem with the cluster; where in-fact, it could be a buggy problem in our code..maybe parsing the output, or something else weird with `df` on the system.

I feel it would be bad for the UX to show a warning/error in the case nothing is _actually_ wrong which is why I opted just for the log message originally, so that this message could be picked up by us, or some automated log analysis. I don't know what the best solution for the user would be in this situation. I just don't want to show a scary orange warning in the UI when nothing is actually wrong.

Thoughts?

Reply:
I see, I was originally anticipate that the warning would mostly be environment issues not a bug in our code. If the warning message is something actionable to the admin then we should show it on UI for them. If we are not confident that this feature will work consistently then let's not include it for now. Let's revisit it some time later once we are more confident on the quality and update the UI if we see specific types of issues that are useful to the admin users.

Reply:
I'm confident that this feature will work for our supported platforms (e.g. Linux -- CentOS 6+ and Ubuntu/Debian-based distros/containers). It's really impossible to test every environment this will be executed in. So if some distro isn't POSIX compliant with the `df` output spec, or maybe a transient error occurs executing `df` (potentially failed syscall, other system issue, etc), or maybe someone is using a home-baked container in k8s/docker using some weird base image is the case that I expect us to reach this particular `return`.

Those potentially transient errors or un-comformity to the POSIX specification is what would result in us hitting this point. I am confident in the logic which parses the output that _is_ POSIX-confirming. Since we can't possibly determine all scenarios which these errors might occur, and the fact that those errors may not be indicative of an issue with the system, which is why I am weary about showing a warning in the UI which may not necessitate any user action (or even be actionable by the user). If you think it's best to just put all the warnings in the UI regardless of the type of error that's ok. I just wanted to have a discussion since it _does_ affect the UX and isn't just some internal log message that we will look for in the future.

Reply:
Makes sense. Let's aim to improve the compatibility as usage grows and add UI warnings case by case if we find them useful.